scheduling/task_run.go

```diff
@@ -0,0 +1,149 @@
+	dummyTrigger  TriggerType = "DummyTrigger"
+)
+
+type Trigger interface {
+	Type() TriggerType
+	Name() string
+}
+
+type OneOffTrigger struct {
+	TriggerName string      `json:"triggerName"`
+	TriggerType TriggerType `json:"triggerType"`
+}
+
+func (t OneOffTrigger) Type() TriggerType {
+	return t.TriggerType
+}
+
+func (t OneOffTrigger) Name() string {
+	return t.TriggerName
+}
+
+type DummyTrigger struct {
+	TriggerName string      `json:"triggerName"`
+	TriggerType TriggerType `json:"triggerType"`
+	DummyField  bool        `json:"dummyField"`
+}
+
+func (t DummyTrigger) Type() TriggerType {
+	return t.TriggerType
+}
+
+func (t DummyTrigger) Name() string {
+	return t.TriggerName
+}
+
+type TaskRunMetadata struct {
+	ID        TaskRunId `json:"runId"`
+	TaskId    TaskId    `json:"taskId"`
+	Name      string    `json:"name"`
+	Trigger   Trigger   `json:"trigger"`
+	Status    Status    `json:"status"`
+	StartTime time.Time `json:"startTime"`
+	EndTime   time.Time `json:"endTime"`
+	Logs      []string  `json:"logs"`
+	Error     string    `json:"error"`
+}
+
+// Formatting
+func (t *TaskRunMetadata) Marshal() ([]byte, error) {
+	return json.Marshal(t)
+}
+func (t *TaskRunMetadata) Unmarshal(data []byte) error {
+	type tempConfig struct {
+		ID        TaskRunId       `json:"runId"`
+		TaskId    TaskId          `json:"taskId"`
+		Name      string          `json:"name"`
+		Trigger   json.RawMessage `json:"trigger"`
+		Status    Status          `json:"status"`
+		StartTime time.Time       `json:"startTime"`
+		EndTime   time.Time       `json:"endTime"`
+		Logs      []string        `json:"logs"`
+		Error     string          `json:"error"`
+	}
+
+	var temp tempConfig
+	if err := json.Unmarshal(data, &temp); err != nil {
+		return fmt.Errorf("failed to deserialize task run metadata: %w", err)
+	}
+
+	if temp.ID == 0 {
+		return fmt.Errorf("task run metadata is missing ID")
+	}
+	t.ID = temp.ID
+
+	if temp.TaskId == 0 {
+		return fmt.Errorf("task run metadata is missing RunID")
+	}
+	t.TaskId = temp.TaskId
+
+	if temp.Name == "" {
+		return fmt.Errorf("task run metadata is missing required fields")
+	}
+	t.Name = temp.Name
+
+	if temp.Status == "" || (temp.Status != Success && temp.Status != Failed && temp.Status != Pending) {
+		return fmt.Errorf("unknown status: %s", temp.Status)
+	}
+	t.Status = temp.Status
+
+	if temp.StartTime.IsZero() {
+		return fmt.Errorf("task run metadata is missing Start Time")
+	}
+	t.StartTime = temp.StartTime
+
+	if temp.EndTime.IsZero() {
+		return fmt.Errorf("task run metadata is missing End Time")
+	}
+	t.EndTime = temp.EndTime
+	t.Logs = temp.Logs
+	t.Error = temp.Error
+
+	triggerMap := make(map[string]interface{})
+	if err := json.Unmarshal(temp.Trigger, &triggerMap); err != nil {
+		return fmt.Errorf("failed to deserialize trigger data: %w", err)
+	}
+
+	if _, ok := triggerMap["triggerType"]; !ok {
+		return fmt.Errorf("target type is missing")
+	}
+
+	if triggerMap["triggerType"] == "OneOffTrigger" {
```

Comment:
Something we should do that i realized from the other one is use the enum here instead of a single string and make it a switch rather than an if

If we can change this for tasks defs too thatd be good